code/datums/components/material/material_container.dm

```diff
@@ -119,27 +107,137 @@
+//======================================LOW LEVEL=========================================
+/**
+ * Inserts the relevant materials from an item into this material container.
+ * This low level proc should not be used directly by anyone
+ *
+ * Arguments:
+ * - [source][/obj/item]: The source of the materials we are inserting.
+ * - multiplier: The multiplier for the materials extract from this item being inserted.
+ * - breakdown_flags: The breakdown bitflags that will be used to retrieve the materials from the source
+ */
+/datum/component/material_container/proc/insert_item_materials(obj/item/source, multiplier = 1, breakdown_flags = mat_container_flags)
+	var/primary_mat
+	var/max_mat_value = 0
 
-/// Proc that allows players to fill the parent with mats
-/datum/component/material_container/proc/on_attackby(datum/source, obj/item/weapon, mob/living/user)
-	SIGNAL_HANDLER
+	var/list/item_materials = source.get_material_composition(breakdown_flags)
+	for(var/MAT in item_materials)
+		if(!can_hold_material(MAT))
+			continue
+		materials[MAT] += OPTIMAL_COST(item_materials[MAT] * multiplier)
+		if(item_materials[MAT] > max_mat_value)
+			max_mat_value = item_materials[MAT]
+			primary_mat = MAT
 
-	user_insert(weapon, user)
+	return primary_mat
+//===================================================================================
 
-	return COMPONENT_NO_AFTERATTACK
 
+//===============================MID LEVEL===================================================
+/**
+ * For inserting an amount of material. Use this to add materials to the container directly
+ *
+ * Arguments:
+ * - amt: amount of said material to insert
+ * - mat: the material type to insert
+ */
+/datum/component/material_container/proc/insert_amount_mat(amt, datum/material/mat)
+	if(amt <= 0)
+		return 0
+	amt = OPTIMAL_COST(amt)
+	if(!has_space(amt))
+		return 0
+
+	var/total_amount_saved = total_amount()
+	if(mat)
+		if(!istype(mat))
+			mat = GET_MATERIAL_REF(mat)
+		materials[mat] += amt
+	else
+		var/num_materials = length(materials)
+		if(!num_materials)
+			return 0
+
+		amt /= num_materials
+		for(var/i in materials)
+			materials[i] += amt
+	return (total_amount() - total_amount_saved)
+
+/**
+ * Proc specifically for inserting items, use this when you want to insert any item into the container
+ * this bypasses most of the material flag checks so much be used by machines like recycler, stacking machine etc that
+ * does not care for such checks
+ *
+ * Arguments:
+ * - [weapon][obj/item]: the item you are trying to insert
+ * - multiplier: The multiplier for the materials being inserted
+ * - breakdown_flags: The breakdown bitflags that will be used to retrieve the materials from the source
+ */
+/datum/component/material_container/proc/insert_item(obj/item/weapon, multiplier = 1, breakdown_flags = mat_container_flags)
+	if(QDELETED(weapon))
+		return MATERIAL_INSERT_ITEM_NO_MATS
+	multiplier = CEILING(multiplier, 0.01)
+
+	var/obj/item/target = weapon
```

Comment:
You're not typecasting target or weapon to anything different, so you could reasonably use weapon for all target instances or rename weapon to target, either or.

Reply:
This isn't for typecasting reasons. target holds the item we are trying to insert and it differs from weapon when the stack is split if there isn't enough space
https://github.com/tgstation/tgstation/blob/b3f22e0a7b90ac12217a9382ea2308a6bb1314cf/code/datums/components/material/material_container.dm#L211
Didn't want to confuse it with weapon for clarity sake